### ISSUE TITLE
[UWP] ILC.exe fails with "MSBUILD : error : Internal compiler error: Sequence contains no elements"

### ISSUE THREAD

**Issue by llfab** [dotnet/runtime#64801]
### Description

I have an UWP application that uses my own .net standard 2.0 assembly as a reference and fails when building with:

Processing application code
Task "RunILTransforms"
Running 'DumpNativeResources' transform, heap at 64 MB.
Running 'Analysis' transform, heap at 78 MB.
Loading 109 modules...
Internal compiler error: Sequence contains no elements
Done executing task "RunILTransforms" -- FAILED.
Done building target "RunILTransforms" in file "ILTransforms" -- FAILED.
Done building project "ILTransforms" -- FAILED.
Compilation failed

I have run ilc.exe manually an enabled all logging and intermediate info. Even with that I cannot see in which exact module the above step fails. 

"Sequence contains no elements" may have to do with LINQ. There is a known issue (https://github.com/scottksmith95/LINQKit/issues/47) with LINQKit. However, this is not used in my assembly.

Yes, there must be something with my own assembly (which has some 100K lines of code) but I would like to investigate which portion of the assembly causes this failure. 

I have run ilc.exe in Visual Studio Debugger (Open file or project -> ilc.exe, provide cmd arguments) but that also just ends without any extra info.

Could you guide me to better analysis options.

### Reproduction Steps

When I build with using native toolchain it works if I do not reference my own assembly. When I reference my own .net standard 2.0 assembly and build the project without native tool chain, everything works. If I build it with native toolchain (which cannot be avoided with ARM64) the build failure appears in the native tool chain (ilc.exe).

### Expected behavior

The app should build using native tool chain

### Actual behavior

The build fails using native tool chain with this full output:

Starting .NET Native compilation

Task "Gatekeeper"
Launching 'C:\Users\FHuegle\.nuget\packages\runtime.win10-x64.microsoft.net.native.compiler\2.2.11-rel-30601-02\tools\x64\ilc\ilc.exe --gatekeeper @"D:\Workspaces\GitLab\Repos\BitsOfNature\ResearchApplications\PelvisSiLiveMr\obj\x64\Release\ilc\intermediate\gkargs.rsp"'
Microsoft (R) .NET Native Gatekeeper Version 2.2.30601.02 
Copyright (C) Microsoft Corporation.  All rights reserved. 
 
Checking assembly 'D:\Workspaces\GitLab\Repos\BitsOfNature\ResearchApplications\PelvisSiLiveMr\obj\x64\Release\ilc\in\BitsOfNature.Core.dll'. 
Checking assembly 'D:\Workspaces\GitLab\Repos\BitsOfNature\ResearchApplications\PelvisSiLiveMr\obj\x64\Release\ilc\in\PelvisSiLiveMr.exe'. 
Checking assembly 'D:\Workspaces\GitLab\Repos\BitsOfNature\ResearchApplications\PelvisSiLiveMr\obj\x64\Release\ilc\in\StereoKit.dll'. 
'C:\Users\FHuegle\.nuget\packages\runtime.win10-x64.microsoft.net.native.compiler\2.2.11-rel-30601-02\tools\x64\ilc\ilc.exe --gatekeeper @"D:\Workspaces\GitLab\Repos\BitsOfNature\ResearchApplications\PelvisSiLiveMr\obj\x64\Release\ilc\intermediate\gkargs.rsp"' completed.
Process execution statistics (all times in milliseconds)
-----------------------------------------------------------------------------
launch  |  start   |   stop   |   end    | duration | lifetime | [slot] name
-----------------------------------------------------------------------------
       0 |      109 |      328 |      390 |      219 |      390 | [0] Gatekeeper
Number of commands:      1
Total system time:       390 ms
System time per command: 390 ms
Total duration:          219 ms
Total lifetime:          390 ms
Requested DOP:           16
Duration DOP:            0.6
Lifetime DOP:            1.0
Done executing task "Gatekeeper".
Done building target "RunGatekeeper" in file "Preparation".

Done building target "CreateToolchainGeneratedAssembly" in file "Preparation".

Task "PreMergeAssemblyClosure"
Done executing task "PreMergeAssemblyClosure".
Done building target "ConstructPreMergeAssemblyClosure" in file "Preparation".

Task "McgReferenceResolveFilePath"
Done executing task "McgReferenceResolveFilePath".
Done building target "PreResolveTransforms" in file "Preparation".

Task "ResolveToolchainRequiredAssemblyRoots"
Done executing task "ResolveToolchainRequiredAssemblyRoots".
Task "ILTransformAssemblyClosure"
Done executing task "ILTransformAssemblyClosure".
Task "ResolveToolchainRequiredAssemblyRoots"
Done executing task "ResolveToolchainRequiredAssemblyRoots".
Task "ReducedCopyAssemblyClosure"
Done executing task "ReducedCopyAssemblyClosure".
Done building target "ConstructILTransformAssemblyClosure" in file "Preparation".
Done building project "Preparation".

Processing application code
Task "RunILTransforms"
Running 'DumpNativeResources' transform, heap at 64 MB.
Running 'Analysis' transform, heap at 78 MB.
Loading 109 modules...
Internal compiler error: Sequence contains no elements
Done executing task "RunILTransforms" -- FAILED.
Done building target "RunILTransforms" in file "ILTransforms" -- FAILED.
Done building project "ILTransforms" -- FAILED.
Compilation failed

### Configuration

Visual Studio 2022
Microsoft.NETCore.UniversalWindowsPlatform 6.2.13
NETStandard.Library 2.0.3

**Comment by llfab** [comment]
Thanks to @tommcdon I was able to fix:
- the failure in ilc run was not when loading any framework assemblies but loading my own assembly
- The assembly (when running in net standard 2.0 mode) implemented some framework classes that are only available in > standard 2.0 just to make the assembly build in 2.0 with not-impl-ex in the mocked functions. The mock implementation was placed in Microsoft.CSharp namespace just like the original class/function. This made ilc assume that this is the very original class from the Microsoft.CSharp assembly failing when searching for methods that are not implemented in the mock classes (and not even used in the code).
- Putting the very same mock classes in an assembly local namespace prevented ilc from assuming it were the original assembly and made ilc run sucessful.

The analysis problem however remains. If ilc fails there seems no option to analyze yourself. I had switched to the msbuild logger (above output) to at least see something. But even in diag level there is no output that would even be close to the failure.

I had tried to build my own ilc from the repo but that is not as easy as building the solution in VS.
